src/amo/reducers/users.js

```diff
@@ -475,15 +518,99 @@ export const addUserToState = ({
   return { byID, byUsername };
 };
 
+type FinishUnsubscribeNotificationParams = {|
+  hash: string,
+  notification: string,
+  token: string,
+|};
+
+export type FinishUnsubscribeNotificationAction = {|
+  type: typeof FINISH_UNSUBSCRIBE_NOTIFICATION,
+  payload: FinishUnsubscribeNotificationParams,
+|};
+
+export const finishUnsubscribeNotification = ({
+  hash,
+  notification,
+  token,
+}: FinishUnsubscribeNotificationParams): FinishUnsubscribeNotificationAction => {
+  invariant(hash, 'hash is required');
+  invariant(notification, 'notification is required');
+  invariant(token, 'token is required');
+
+  return {
+    type: FINISH_UNSUBSCRIBE_NOTIFICATION,
+    payload: { hash, notification, token },
+  };
+};
+
+type AbortUnsubscribeNotificationParams = {|
+  hash: string,
+  notification: string,
+  token: string,
+|};
+
+export type AbortUnsubscribeNotificationAction = {|
+  type: typeof ABORT_UNSUBSCRIBE_NOTIFICATION,
+  payload: AbortUnsubscribeNotificationParams,
+|};
+
+export const abortUnsubscribeNotification = ({
+  hash,
+  notification,
+  token,
+}: AbortUnsubscribeNotificationParams): AbortUnsubscribeNotificationAction => {
+  invariant(hash, 'hash is required');
+  invariant(notification, 'notification is required');
+  invariant(token, 'token is required');
+
+  return {
+    type: ABORT_UNSUBSCRIBE_NOTIFICATION,
+    payload: { hash, notification, token },
+  };
+};
+
+type GetUnsubscribeKeyParams = {|
+  hash: string,
+  notification: string,
+  token: string,
+|};
+
+export const getUnsubscribeKey = ({
+  hash,
+  notification,
+  token,
+}: GetUnsubscribeKeyParams) => {
+  invariant(hash, 'hash is required');
+  invariant(notification, 'notification is required');
+  invariant(token, 'token is required');
+
+  return `${hash}-${notification}-${token}`;
+};
+
+export const isUnsubscribedFor = (
+  usersState: UsersState,
+  hash: string,
+  notification: string,
+  token: string,
+) => {
+  return usersState.isUnsubscribedFor[
+    getUnsubscribeKey({ hash, notification, token })
+  ];
+};
+
 type Action =
+  | AbortUnsubscribeNotificationAction
   | FetchUserAccountAction
   | FetchUserNotificationsAction
   | FinishUpdateUserAccountAction
-  | UpdateUserAccountAction
+  | FinishUpdateUserAccountAction
```

Comment:
This appears to be repeated, whereas `FinishUnsubscribeNotificationAction` does not appear in this list. I wonder why this is not generating any errors?

Reply:
oh oops. I think we need a special Flow notation to get an error in the reducer.